### DISCUSSION TITLE
Null-conditional operators and invoking a delegate

### DISCUSSION THREAD

**Discussion by DavidArno** [dotnet/csharplang#4725] [Language Ideas] [answered]
It strikes me that there is a missing scenario for null conditional operators: invoking a delegate.

I can invoke a delegate in one of two ways:
```cs
d.Invoke();
d();
```
And I can mix this with the null conditional operator:
```cs
d?.Invoke();
```
Indexers have their own null conditional operator:
```cs
y = x?[i];
```
So it seems natural to assume there would be a conditional operator for the shorthand form of delegate invocation. But the obvious form:
```cs
d?();
```
has issues as it's not until `;` that the compiler could know that this is a conditional operator/delegate invocation mix and not the start of a ternary expression. The alternative:
```cs
d?.();
```
looks a bit odd, but maybe that's OK.

Anyway, I've no idea if this topic has already been addressed, is in consideration or if there's a better solution that I haven't thought of. So I'm just throwing these thoughts out into the open to see what the response is.

**Comment by canton7** [comment]
Related: #3257 (although I don't see the syntax `?.(...)` there, so I'm not calling it a duplicate)

**Reply by DavidArno** [reply]
Good point. Is a slightly different syntax grounds for it not being a duplicate though? There again, it's been closed presumably due to the focus on how that proposed syntax wouldn't work. So I'll leave this open for a bit longer even though I suspect the conversation has been had already.

**Comment by HaloFour** [comment, marked as answer]
For context:

Archived meeting notes from Feb 3, 2014: [Wayback Machine -> Roslyn on CodePlex](https://web.archive.org/web/20140426114809mp_/http://roslyn.codeplex.com/discussions/540281)

> ### Which operators does it combine with?
> The main usage of course is with dot, as in `x?.y` and `x?.m(…)`. It also potentially makes sense for element access `x?[…]` and invocation `x?(…)`. And we also have to consider interaction with indexed names, as in `x?.$y`.
> 
> We’ll do element access and indexed member access, but not invocation. The former two make sense in the context that lightweight dynamic is addressing. Invocation seems borderline ambiguous from a syntactic standpoint, and for delegates you can always get to it by explicitly calling Invoke, as in `d?.Invoke(…)`.

**Reply by DavidArno** [reply]
Thanks, It seems safe to assume that the team's position hasn't changed, so I'll mark this as the answer.

**Reply by CyrusNajmabadi** [reply]
The ambiguity is the biggest issue for me.  `  x ? (...` is already teh start of many reasonable constructs.  Disambiguating here would be a PITA.

**Reply by canton7** [reply]
The OP calls that out? They suggest `?.(...)`, which I don't think I've seen proposed before (not that I'm necessarily in favour of it).

**Reply by CyrusNajmabadi** [reply]
Sure.  Then it falls into: just not important enough, and the existing solution is sufficient :)
